fml/platform/darwin/scoped_nsobject.h

```diff
@@ -37,29 +37,49 @@ namespace fml {
 // We check for bad uses of scoped_nsobject and NSAutoreleasePool at compile
 // time with a template specialization (see below).
 
+#if __has_feature(objc_arc)
+class scoped_nsprotocol_memory_management {
+ public:
+  static id Retain(__unsafe_unretained id object) { return object; }
+  static id Autorelease(__unsafe_unretained id object) { return object; }
+  static void Release(__unsafe_unretained id object) {}
+  static id InvalidValue() { return nil; }
+};
+#else
+class scoped_nsprotocol_memory_management {
+ public:
+  static id Retain(id object) { return [object retain]; }
+  static id Autorelease(id object) { return [object autorelease]; }
+  static void Release(id object) { [object release]; }
+  static id InvalidValue() { return nil; }
+};
+#endif
```

Comment:
I believe this introduces another ODR violation. You again have symbols with identical names where the definitions are different in different translation units (ARC vs non-ARC). In this case, the violation is even clearer since the definitions consist of different tokens, rather than the earlier iteration where the same tokens led to different machine code due to ARC.

Reply:
What symbols do you believe are the same between the ARC and non-ARC TUs?  If `fml::scoped_nsobject<Foobar*> foo;` is compiled with ARC we should have a `fml::scoped_nsobject<Foober* __strong>::scoped_nsobject()` symbol and if it is compiled without ARC it will be a `fml::scoped_nsobject<Foobar*>::scoped_nsobject()` symbol.

I don't think there is a way, and it definitely isn't easy, to create a `fml::scoped_nsobject<Foobar*>::scoped_nsobject();` when compiling with ARC.

Reply:
The symbols in the `scoped_nsprotocol_memory_management` class.

Reply:
Though I suppose they may also end up with different mangling due to taking parameters / returning values that are Obj-C types? I'm not clear if the difference in mangling is specific to template type parameters, or if it extends to parameters and return types as well.

Reply:
Ahh good catch, thanks!  I addressed it by making scoped_nsprotocol_memory_management templated around the NST.

Reply:
Ahh, yea the return type would have been `id` versus `id strong`.  Return types are not codified in symbol names in c++, so it would have been a problem for the InvalidValue.  This is just safer to template them.

Reply:
(at the cost of a slightly larger binary)

Reply:
> What symbols do you believe are the same between the ARC and non-ARC TUs? If `fml::scoped_nsobject<Foobar*> foo;` is compiled with ARC we should have a `fml::scoped_nsobject<Foober* __strong>::scoped_nsobject()` symbol and if it is compiled without ARC it will be a `fml::scoped_nsobject<Foobar*>::scoped_nsobject()` symbol.

If `scoped_nsobject` evaluates to different classes in ARC and non-ARC, then you're just pushing the ODR violation up a level; any C++ class with a scoped_nsobject ivar that's included from both ARC and non-ARC code (which is the case this whole effort is intended to allow, so not an edge case) is now violating the ODR, because there are two different versions of that class.

Reply:
@stuartmorgan I think you're mistaken.  If you have:
```c++
class Foo {
 public:
  Foo();
 private:
  fml::scoped_nsobject<NSArray*> array_;
};
```

You will only get 1 version of `Foo::Foo()`, it will be the one that gets compiled from `foo.mm`.  It will reference `fml::scoped_nsobject<NSArray*>::scoped_nsobject()` or `fml::scoped_nsobject<NSArray* __strong>::scoped_nsobject()` depending on if it was compiled with ARC or not.  Were you thinking of a different setup?

Reply:
Stuart's correct. Consider what happens if an inline function in a header file interacts with a `scoped_nsobject<NSArray *>`. It'll emit references to different symbols depending on whether it is imported into an ARC or MRC translation unit. That is an ODR violation since the same function now has different definitions in different translation units.

Reply:
Ahh yea.  In the above example `~Foo()` would be problematic, right?

One possible way forward would be to annotate the code correctly and prove that the machine code is equivalent (if not equal) with unit tests.  That sounds like a fair bit of work and has a risk of not panning out.

Another would be do so something similar to what @bdash mentioned where we use `void*` or` CFTypeRef` to bypass ARC's code generation to make the machine code equivalent.

I'm loath to take any more of your time, I thought I could hammer this out quickly.  We can live without ARC, it's just a shame to live without it.  I've spent a lot of time fixing memory bugs in our objc code and we have a whole lot more to write while we catch up with unit tests.